Allow org.openssl.engine: for each key input option (for apps supporting -engine)

DDvO opened the pull request:
The original aim of this PR was to fix a regression that happened during the CMP contribution in `cmp.c`
on the usability of the optional value prefix `engine:` for the `-newkey`, -`key`, and `-tls_key` options 
when used with the `-engine` option.

Meanwhile it is about allowing the prefix `org.openssl.engine:` to flag engine-based keys per key input option
also for other CLI apps.

levitte commented:
Uhmmm, have we ever had support for `engine:`?
I recall trying to promote that kind of uri, and got a lot of pushback because of an unregistered scheme....

DDvO commented:
> Uhmmm, have we ever had support for `engine:`?

I just searched through the commit logs but could not find `engine:` (nor `eng:`) there.

> I recall trying to promote that kind of uri, 

Pleased to hear.

>and got a lot of pushback because of an unregistered scheme....

Everything new / inventive is likely unregistered ;-)

The charm of such a prefix is that - in contrast to the `-keyform engine`, `-inform engine` etc. options provided by many apps (including cmp) is that it can be given directly with the respective key input, without the inconvenience to use a separate option. 
Moreover, when there is more than one key input option their relation with key form options is error-prone.

So why not simply allow it (for backward compatibility, in addition to the existing `-keyform` etc.) options.
I can add documentation for `engine:` (as already done for `cmp`) also for the other apps.

DDvO commented:
I've just rebased this and moved the fixes not directly pertaining to the engine regression to an independent PR: #13409.

richsalz commented:
Don't add a new URI scheme.  The existing options work fine, or should be fixed.

DDvO commented:
> Don't add a new URI scheme. The existing options work fine, or should be fixed.

@richsalz I wonder why you think one should not add such a scheme.
As explained, the existing ones are a bit awkward to use.

The `engine:` scheme IMHO fits well with related ones like `file:` and `pkcs11:`
- see for instance https://tools.ietf.org/html/rfc7512#section-3

Note also that the `engine:` prefix is already documented within `cmp.c` and `openssl-cmp.pod.in`,
and it just does not work, so the primary aim of this PR is to fix that for the CMP app.

On this occasion I'd say it's reasonable to offer this possibility also to other apps 
by implementing it in `apps.c`, as done here in a straightforward way.

richsalz commented:
I am opposed to this because, frankly, this is going to end up being a partial solution so that one command works differently and the others do not.

levitte commented:
No, @richsalz. If you read again what @DDvO wrote, he intends this as a general thing that works automatically with all key options in all our apps, without the need for a separate form option.

levitte commented:
BTW, I can try and resurrect what I did. It was a bit more involved, with an app specific OSSL_STORE loader, which meant that the uris could be passed to SSL functions, X509_STORE functions, etc etc etc.

richsalz commented:
I don't see manpage changes to document this new engine URI, which is why I am skeptical of it being universal.

DDvO commented:
> I don't see manpage changes to document this new engine URI, which is why I am skeptical of it being universal.

@richsalz, as I mentioned above:

> I can add documentation for `engine:` (as already done for `cmp`) also for the other apps.

@levitte, shall I just do that?

levitte commented:
> > I can add documentation for `engine:` (as already done for `cmp`) also for the other apps.
> 
> @levitte, shall I just do that?

If this is made into a global thing, then it may be a good idea to have its main documentation in `doc/man1/openssl.pod`, and refer to that from the other pages.  BTW, it may be that the current `Format options` in section `doc/man1/openssl.pod` needs a thorough cleanup...  I only read it very quickly, so I might be wrong, please have another look.

DDvO commented:
I'll do that soon...

richsalz commented:
And when/how is it used by, say, x509 etc?

levitte commented:
> And when/how is it used by, say, x509 etc?

@richsalz, I believe you know "grep"?

``` console
$ git grep -E 'load_(pub)?key' apps/x509.c
apps/x509.c:        fkey = load_pubkey(fkeyfile, keyformat, 0, NULL, e, "forced key");
apps/x509.c:                    Upkey = load_key(keyfile, keyformat, 0,
apps/x509.c:                    CApkey = load_key(CAkeyfile, CAkeyformat,
apps/x509.c:                    pk = load_key(keyfile, keyformat, 0,
```

mattcaswell commented:
I thought the objection before was that "engine" is not registered here:

https://www.iana.org/assignments/uri-schemes/uri-schemes.xhtml

DDvO commented:
> I thought the objection before was that "engine" is not registered here:
> https://www.iana.org/assignments/uri-schemes/uri-schemes.xhtml

I still don't see that it would be important to register there all schemes we use, but maybe it's helpful?
How much effort would it take to do that for `engine:`, and how long could that take?
If registration is actually considered relevant, I think we could also do that _after_ implementing it here.

levitte commented:
RFC 7595 does indeed say that even private schemes with "global" looking names should be registered as "provisional", so I can understand why some are pushing for use to register `engine:`...  personally, I've concerns of another nature, in that ENGINE is a deprecated interface, subject for removal in a small number of years, so I have no desire to make the URI scheme that level of permanent.

That leads to an interesting bit in [Section 3.8 of RFC 7595](https://tools.ietf.org/html/rfc7595#section-3.8), which discusses scheme names, and most specifically this paragraph:

>    Some organizations desire their own namespace for URI scheme names
>    for private use (see Section 6).  In doing so, it is important to
>    prevent collisions and to make it possible to identify the owner of a
>    private-use scheme.  To accomplish these two goals, such
>    organizations SHOULD use a prefix based on their domain name,
>    expressed in reverse order.  For example, a URI scheme name of
>    com.example.mything might be used by the organization that owns the
>    example.com domain name.  Care must be taken, however, if the
>    organization later loses the domain name embedded in their scheme
>    names since domain name registrations are not permanent.  To
>    associate the private-use scheme name with the original organization,
>    the private-use scheme can be registered using the registration
>    procedure in Section 7.

While this RFC still *strongly encourages* registration, it appears to me that it's not actually mandatory, and that a URI scheme such as `org.openssl.engine:` would at least have us escape the global URI requirements.

DDvO commented:
Interesting solution! 
Though I would have preferred `engine:` for convenience (or at least `ossl_engine:` for brevity),
I'd be fine with using `org.openssl.engine:` here to be on the safe side.

This is still much more straightforward and simpler to use than combinations such as `-keyform engine` ... `-key <id>` , 
in particular if (as typical with CMP) an app invocation uses several private keys where part of them are engine-based.

richsalz commented:
@levitte my concern is not the calls (yes, thanks for the grep usage) but the documentation.

levitte commented:
> @levitte my concern is not the calls (yes, thanks for the grep usage) but the documentation.

So all the scepticism that was put on display was really just a "please make sure this is properly documented!" ?

richsalz commented:
I dislike your minimizing my concern.  The fact that others were addressed doesn't mean that the overall concern isn't well-founded, especially as there is a history here of "I'll just fix CMP"

levitte commented:
> I dislike your minimizing my concern.

My apologies, that wasn't what I was trying, even though I can see it came off that way.  And you're right, the PR title doesn't quite reflect that this has a broader impact.

DDvO commented:
I've just updated the PR title to reflect the current state of the discussion.

richsalz commented:
Thank you.

DDvO commented:
> especially as there is a history here of "I'll just fix CMP"

@richsalz, indeed it happened many times that I added/extended some CMP feature or fixed some problem with the CMP implementation and these improvements had an impact also on other parts of the project.

Just think of my numerous small and partly very major PRs such as the complete overhaul of the HTTP code, various improvements of file (in particular, PKCS#12) loading, fixes of certificate handling, enhancements of the test machinery, extensions/improvements of documentation and of various tools.

In most such cases the general nature of those improvements was deliberate from their beginning.
In this case the generalization of OpenSSL apps besides `cmp` w.r.t referencing engine-based keys was more accidental, but IMHO it is going to be a pretty positive general side-effect.

richsalz commented:
@ddvo when I said things are done partially, leaving complete changes to another PR, it wasn't just about your work on CMP. For example, the two sets of deprecation macros we have now.

DDvO commented:
@levitte wrote:

> If this is made into a global thing, then it may be a good idea to have its main documentation in doc/man1/openssl.pod, and
> refer to that from the other pages. BTW, it may be that the current Format options in section doc/man1/openssl.pod needs a 
> thorough cleanup... I only read it very quickly, so I might be wrong, please have another look.

I've just done this, so the `org.openssl.engine:` prefix is now documented centrally in `openssl.pod`.
I went though the `.pod.in` files of all apps having key(form) options and updated and cleaned up their documentation.

I also cleaned up the `Format options` section of `openssl.pod`. For whatever reasons, it had mentioned some `-input` option (text added by commit 777182a0c77ee374e43b94546f49b25f37945c0e) that is nowhere to find in the sources nor elsewhere in the documentation, so I removed that spurious thing.

DDvO commented:
I've just rebased this PR to solve a trivial merge conflict.
Anything more/else to do here?

kaduk commented:
> I still don't see that it would be important to register there all schemes we use, but maybe it's helpful?
How much effort would it take to do that for engine:, and how long could that take?

Sorry, I'm quite behind on openssl mail during IETF meeting week and didn't see this sooner.

(Yes, I was/am one of the people complaining loudly about using an unregistered URI scheme.)

It is actually pretty straightforward to register a provisional URI scheme, and the documentation need not be particularly extensive/comprehensive.  It can go as quickly as a few weeks, and there is a community of experienced parties willing to offer advice on how to improve the specification, during that period.

DDvO commented:
> > I still don't see that it would be important to register there all schemes we use, but maybe it's helpful?
> > How much effort would it take to do that for `engine:`, and how long could that take?
> 
> It is actually pretty straightforward to register a provisional URI scheme, and the documentation need not be particularly extensive/comprehensive. It can go as quickly as a few weeks, and there is a community of experienced parties willing to offer advice on how to improve the specification, during that period.

Thanks @kaduk for answering this.

So it would be doable to register the `engine:` scheme with some (limited) effort in reasonable time.

The obvious follow-up question now is whether we decide to go for that and who can and is willing to pursue this,
or if we stick (at least for the time being) to the private scheme `org.openssl.engine:` currently proposed here?

mattcaswell commented:
> It is actually pretty straightforward to register a provisional URI scheme, and the documentation need not be particularly extensive/comprehensive.

@kaduk can you give us some pointers on what we need to do and where we would start with this?

DDvO commented:
Ping @kaduk

kaduk commented:
Sorry for the delay -- with the US holiday and the thing I thought would take a couple hours actually taking a couple days, I didn't have much time for openssl this week.

The registration policy is "first come first served" for provisional registrations, so technically we can just ask for the name, but it is strongly encouraged that we have some written specification for the scheme definition (e.g., an Internet-Draft), including security considerations and any deviation from the guidelines for permanent registrations.  This is covered in https://tools.ietf.org/html/rfc7595#section-4 as modified by https://www.rfc-editor.org/errata_search.php?eid=4420 .  (The references for the registry itself, https://www.iana.org/assignments/uri-schemes/uri-schemes.xhtml , include RFC 8615 as well, that reminds us to consider whether we wish to carve out a "well-known" portion of the namespace.)

I'm happy to take the lead on getting this stuff written up in the form of an I-D, but would need some guidance on what we actually want the semantics of an "engine:" URI scheme to be.  (Is it just a single flat/non-hierarchical name?)

DDvO commented:
Thanks a lot @kaduk for providing those details on URI scheme registration.

> want the semantics of an "engine:" URI scheme to be. (Is it just a single flat/non-hierarchical name?)

The intended use is in the context of referencing a (private) key for use with crypto operations such as signing.
The semantics is that the key reference is to be interpreted relative to some crypto engine.

For example, in a CLI option `-signkey engine:id_cafe1234`.
The engine (so far) is determined by other means, such as the CLI option `-engine pkcs11`.

It would make sense to (optionally?) include the engine identifier as a sub-name, e.g.: `engine:pkcs11:id_cafe1234`.

kaduk commented:
Looking at the RFC 3986 breakdown of URI into scheme, authority, path, query, and fragment, it seems like the usage we're describing implicity is using localhost as the authority (or lacking an authority component entirely), so the engine name would probably end up in the path.  There is perhaps a bit of a philosophical question of whether the key identifier belongs in the (hierarchical) path or is part of the (non-hierarchical) query component.  IMO there is a hierarchical namespace where the engine name is the first level of hierarchy and the key name the second level, since the interpretation of a given key name is dependent on the particular engine in question.  So that would give us something like `engine:///pkcs1/id_cafe1234` and means that the engine name would always need to be present, with the existing `-engine pkcs11` would be redundant.  It's not clear whether that should be considered a good change or a bad one, though...

DDvO commented:
To me it does make sense to (always) have the engine identifier included because it makes the engine to use explicit.
As mentioned, this also has the advantage to be able to use different engines for different keys, which was not possible with the "global" option `-engine`.
For backward compatibility we could say that `-engine XYZ` has the effect of an implicit prefix `engine:///XYZ/` (to all engine-enabled input options of the given app, such as `-signkey`).

DDvO commented:
BTW, the object identifier following the engine identifier could itself contain (lower-level) schemes.

For instance, lib11 allows besides simple object identifiers such as `id_cafe1234`, `label_mylabel`, `3:4711` and `5fe858bc3f5afb656d1aeb8ca3a1cdeeb4df4016` also structured identifiers 
such as `pkcs11:object=my-private-key;type=private;pin-value=1234` according to https://tools.ietf.org/html/rfc7512.

levitte commented:
`engine:///pkcs11/id_cafe1234` seems incredibly clumsy for a URI scheme that doesn't reasonably have an authority in the web kind of sense.  How about `engine:pkcs11:id_cafe1234`?  It's a perfectly legitimate URI according to [RFC 3986](https://tools.ietf.org/html/rfc3986).  For a loader, it's a piece of cake to extract the ENGINE id (`pkcs11` in this example) from such a URI, and simply consider the rest to be a key id understandable by the ENGINE in question.  This is also more in line with already existing schemes where engines are used (such as [pkcs11:](https://tools.ietf.org/html/rfc7512)).

Also, for a URI scheme that's meant to have a relatively short life span, `engine:` is much too generic, and `org.openssl.engine:` would be preferable.

DDvO commented:
I also would prefer the more straightforward syntax such as `engine:engine-id:engine-based-key-id`.

DDvO commented:
> Also, for a URI scheme that's meant to have a relatively short life span, `engine:` is much too generic, and `org.openssl.engine:` would be preferable.

This touches a fundamental issue: How much effort and time is it really worth spending on officially registering an engine scheme, given that the engine mechanism is already deprecated?
BTW, any estimations on how many years it will still be supported by OpenSSL?

If we just use the "private" scheme `org.openssl.engine:` apparently we don't need to register it, 
and this PR already implements this simple solution.

levitte commented:
> This touches again the fundamental issue: 
> How much effort and time is it really worth spending on officially registering an engine scheme,  in particular given that the engine mechanism is already deprecated?

If @kaduk is willing to put in the effort, then there's at least one that thinks it's worth the effort, and ready to actually do it.  That's good enough for me, I see no reason whatsoever not to take that gift (which is what it is), with all the gratitude.

> BTW, any estimations on how many years it will still be supported by OpenSSL?

Our deprecation policy  says this:

> No existing public interface can be removed until its replacement has been in place in an LTS stable release. The original interface must also have been documented as deprecated for at least 5 years. A public interface is any function, structure or macro declared in a public header file.

Considering we're making ENGINEs deprecated with 3.0, it's at least 5 years from when that version is released.

> If we just use the "private" scheme `org.openssl.engine:` apparently we don't need to register it

No, that's not quite it.  As I said earlier, it's still strongly encouraged, but I imagined that it could be easier to [ahem] sneak away from that encouragement when it was made more private.  @kaduk interfered, though :wink:

levitte commented:
BTW, remember that I talking about resurrecting an actual OSSL_STORE loader that I had originally implemented in #7390?  Coming up!

levitte commented:
See #13570

DDvO commented:
I meanwhile would close this PR because it has been superseded by #13570.

Yet it contains the above discussion with @kaduk et al. on registering the engine scheme.
How to proceed with this?

levitte commented:
@DDvO says:
> How to proceed with this?

I believe that @kaduk asked us for semantic documentation of the URI scheme.
Me, I'm not exactly sure what to say, except with the same thing that's in the recent commits, i.e.:

    org.openssl.engine:{engineid}:{keyid}

Where `{engineid}` is an engine name like `pkcs11`, and `{keyid}` (i.e. "whatever follows") is used as a identity string in calls to any ENGINE function that takes those, i.e. these:

``` C
EVP_PKEY *ENGINE_load_private_key(ENGINE *e, const char *key_id,
                                  UI_METHOD *ui_method, void *callback_data);
EVP_PKEY *ENGINE_load_public_key(ENGINE *e, const char *key_id,
                                 UI_METHOD *ui_method, void *callback_data);
```

Individual engines define what `{keyid}` can actually be.

@kaduk, is that enough, or does that need some more fleshing out?

kaduk commented:
That's enough for me to work with, at least for now.
And there's no reason we can't continue a conversation even if the pull request is marked as closed, should there be a need.
So I will go ahead and close this one out.
